pkg/controller/sparkapplication/sparkapp_metrics.go

```diff
@@ -0,0 +1,229 @@
+		prometheus.CounterOpts{
+			Name: util.CreateValidMetricNameLabel(prefix, "spark_app_success_count"),
+			Help: "Spark App Success Count via the Operator",
+		},
+		labels,
+	)
+	sparkAppFailureCount := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: util.CreateValidMetricNameLabel(prefix, "spark_app_failure_count"),
+			Help: "Spark App Failure Count via the Operator",
+		},
+		labels,
+	)
+	sparkAppSuccessExecutionTime := prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Name: util.CreateValidMetricNameLabel(prefix, "spark_app_success_execution_time_microseconds"),
+			Help: "Spark App Runtime via the Operator",
+		},
+		labels,
+	)
+	sparkAppFailureExecutionTime := prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Name: util.CreateValidMetricNameLabel(prefix, "spark_app_failure_execution_time_microseconds"),
+			Help: "Spark App Runtime via the Operator",
+		},
+		labels,
+	)
+	sparkAppExecutorFailureCount := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: util.CreateValidMetricNameLabel(prefix, "spark_app_executor_failure_count"),
+			Help: "Spark App Failed Executor Count via the Operator",
+		},
+		labels,
+	)
+	sparkAppExecutorSuccessCount := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: util.CreateValidMetricNameLabel(prefix, "spark_app_executor_success_count"),
+			Help: "Spark App Successful Executor Count via the Operator",
+		},
+		labels,
+	)
+	sparkAppRunningCount := util.NewPositiveGauge(util.CreateValidMetricNameLabel(prefix, "spark_app_running_count"),
+		"Spark App Running Count via the Operator", labels)
+	sparkAppExecutorRunningCount := util.NewPositiveGauge(util.CreateValidMetricNameLabel(prefix,
+		"spark_app_executor_running_count"), "Spark App Executor Running Count via the Operator", labels)
+
+	util.RegisterMetric(sparkAppSubmitCount)
+	util.RegisterMetric(sparkAppSuccessCount)
+	util.RegisterMetric(sparkAppFailureCount)
+	util.RegisterMetric(sparkAppSuccessExecutionTime)
+	util.RegisterMetric(sparkAppFailureExecutionTime)
+	util.RegisterMetric(sparkAppExecutorFailureCount)
+	util.RegisterMetric(sparkAppExecutorSuccessCount)
+
+	return &sparkAppMetrics{
+		labels,
+		prefix,
+		sparkAppSubmitCount,
+		sparkAppSuccessCount,
+		sparkAppFailureCount,
+		sparkAppRunningCount,
+		sparkAppSuccessExecutionTime,
+		sparkAppFailureExecutionTime,
+		sparkAppExecutorRunningCount,
+		sparkAppExecutorFailureCount,
+		sparkAppExecutorSuccessCount,
+	}
+}
+
+func fetchMetricLabels(specLabels map[string]string, labels []string) map[string]string {
+	// Transform spec labels since our labels names might be not same as specLabels if we removed invalid characters.
+	validSpecLabels := make(map[string]string)
+	for labelKey, v := range specLabels {
+		newKey := util.CreateValidMetricNameLabel("", labelKey)
+		validSpecLabels[newKey] = v
+	}
+
+	metricLabels := make(map[string]string)
+	for _, label := range labels {
+		metricLabels[label] = validSpecLabels[label]
```

Comment:
It's better to do the following instead:
```
if value, ok := validSpecLabels[label]; ok {
  metricLabels[label] = value
} else {
  metricLabels[label] = "Unknown"
}
```

Reply:
done